State sync engine design doc (#677)
onflow/flow: roadmap/state-sync-protocol.md

* A *session* orchestrates the flow of messages with other peers for discovering and retrieving the individual blocks which comprise a single high-level object.

The full details of the Bitswap protocol can be found [here](https://research.protocol.ai/publications/accelerating-content-routing-with-bitswap-a-multi-path-file-transfer-protocol-in-ipfs-and-filecoin/delarocha2021.pdf).

> To avoid conflicting terminology, we will henceforth refer to blocks in Bitswap as *blobs*.

### State Computation

Execution nodes, upon executing the transactions in each block, will compute the following data:

```protobuf
message ExecutionData {
    repeated Collection collections = 1;
    repeated TransactionResult txResults = 2;
    repeated Event events = 3;
    repeated TrieUpdate trieUpdates = 4;
}
```

They will then serialize each of these fields<sup>[1](#footnotes)</sup> and compute a root CID from the serialized data as follows:

1. Split the data into blobs of up to 1MB (the message size limit of the [underlying networking stack](https://docs.ipfs.io/concepts/libp2p/)).
2. Compute the CID of each blob and store it in the blobstore.

> 3. If the number of CIDs computed in the previous step is 1, then return this CID. Otherwise, concatenate the CIDs and repeat from step 1<sup>[2](#footnotes)</sup>.

Reviewer: Or at least add:

Otherwise, concatenate the CIDs and repeat from step 1 with the new concatinated data as a new blob

# Execution State Synchronization Protocol

## Context

Today, Access nodes act as the gateway to the Flow network, providing a query endpoint for users (REST, gRPC) and performing important admission control duties such as:
* Rate limiting of requests (script executions, Collection requests, account queries, etc.)

> * Basic validation, preventing bad transactions from entering the network

Reviewer 2: Technically, ANs can't _prevent_ bad transactions from entering the network. But they can provide a first level of filtering (imperfect but providing a significant hurdle to spamming the network).

> Execution nodes and Collection nodes currently still expose public APIs, vestiges of an earlier version of the protocol in which Access nodes had not yet been introduced. However, allowing users to bypass the Access layer and make direct requests to these nodes not only dilutes the valuable resources which should be reserved for performing tasks crucial to the protocol (especially in the case of Execution nodes), but it also leaves them at risk of denial-of-service attacks. Going forward, in order to protect the other staked nodes of the network, the Access layer will become the *only* way for users to interact with the network.

Reviewer: I feel like theres some inaccuracies here.

> vestiges of an earlier version of the protocol in which Access nodes had not yet been introduced

I'm not sure if this is the right way to think about this. Even after adding the access layer, we were still adamant about being able to interact with the chain through Collection and Execution nodes, as a feature. This section in the doc seems implies that it was an unwanted side effect.

> the Access layer will become the *only* way for users to interact with the network.

Also, this is quite contentious at the moment, so I'm not so sure it makes sense to have it here before a decision is made.

Reviewer 2: > Execution nodes and Collection nodes currently still expose public APIs, vestiges of an earlier version of the protocol in which Access nodes had not yet been introduced.

Access nodes were already part of the architecture _before_ we started with the Go implementation. The reasons why we started with APIs on the ENs and Collectors was simply a short-cut. 

```suggestion
Execution nodes and Collection nodes currently still expose public APIs, vestiges of earlier short cuts. However, allowing users to bypass the Access layer and make direct requests to these nodes not only dilutes the valuable resources which should be reserved for performing tasks crucial to the protocol (especially in the case of Execution nodes), but it also leaves them at risk of denial-of-service attacks. Going forward, in order to protect the other staked nodes of the network, the Access layer will become the *only* way for users to interact with the network.
```

Reviewer 2: > the Access layer will become the only way for users to interact with the network.

agree with Kan. I think the sync engine will be needed either way, no matter whether the EN or the AN exposes the API. I would argue as follows:

> It is important for network-external entities as well as network-internal nodes to be able to acquire a full copy of the execution state. This is important to remove load from the Execution node. Furthermore, due to their (intentionally) low number, entities other than the Execution Nodes should be able to hold (and potentially archive) the full state to further decentralization.

Author: > Also, this is quite contentious at the moment, so I'm not so sure it makes sense to have it here before a decision is made.

yeah, this was written before we had our previous discussion on Access nodes, and I haven't had time to update it since then

At the moment, the [Access API](/docs/content/access-api.md) contains a few methods in particular that proxy requests behind the scenes to other nodes (possibly from previous sporks):

* [`GetCollectionByID`](/docs/content/access-api.md/#getcollectionbyid) forwards the request to a Collection node and returns the response.
* [`GetTransaction`](/docs/content/access-api.md/#gettransaction) / [`GetTransactionResult`](/docs/content/access-api.md/#gettransactionresult) forwards the request to a Collection / Execution node and returns the response.
* [`ExecuteScriptAtLatestBlock`](/docs/content/access-api.md/#executescriptatlatestblock) / [`ExecuteScriptAtBlockID`](/docs/content/access-api.md/#executescriptatblockid) / [`ExecuteScriptAtBlockHeight`](/docs/content/access-api.md/#executescriptatblockheight) forwards the request to an Execution node and returns the response.
* [`GetAccountAtLatestBlock`](/docs/content/access-api.md/#getaccountatlatestblock) / [`GetAccountAtBlockHeight`](/docs/content/access-api.md/#getaccountatblockheight) forwards the request to an Execution node and returns the response.

> Ideally, Access nodes would be able to serve such requests from their own local state to avoid introducing additional traffic and putting excessive load on nodes within the network. To do this, they need the ability to maintain an up-to-date copy of the execution state and all transaction data.

Reviewer: :+1:

> The [Flow DPS](https://github.com/optakt/flow-dps) has been an important first step in this direction, providing a scalable and efficient way to access the history of the Flow execution state. It reads the protocol state by following block proposals via the [`ConsensusFollower`](https://github.com/onflow/flow-go/blob/master/follower/consensus_follower.go) library, and reads the execution-related data from records written to a Google Cloud Storage bucket by an Execution node. Going forward, we will want to incorporate this functionality into the Access node itself.

Reviewer 2: > The [Flow DPS](https://github.com/optakt/flow-dps) ...

Would suggest to write out the full name at the first occurrence: 

> The [Flow Data Provisioning Service [DPS]](https://github.com/optakt/flow-dps) ...

## Objective

We wish to phase out the GCS bucket architecture and design a state synchronization protocol that does not rely on any third-party cloud providers, yet still enables the scalable transmission of execution state updates and transactions data from Execution nodes to Access nodes. 

Such a protocol should require a minimal amount of additional work from Execution nodes, and in particular it should not involve them doing a fan-out of the state they generate. Instead, Execution nodes should only need to answer a bounded number of initial state requests, after which the further exchange of this data should be delegated among Access nodes themselves. This allows us to leverage the scability of the Access node layer and gain increased data dissemination robustness and decentralization.

## Design

### Data Transfer

The core data exchange protocol we build on top of is called [Bitswap](https://docs.ipfs.io/concepts/bitswap/). At a high level, Bitswap's main responsibilities are to acquire chunks of data requested by the client from the network, and send chunks of data in its possession to other peers who want them.

Bitswap functions with a few fundamental concepts:

> * A *block* represents a raw chunk of data.
> * Each block is uniquely identified by a [*content identifier (CID)*](https://docs.ipfs.io/concepts/content-addressing/), which is based on a cryptographic hash of the block's contents.
> * A *blockstore* is a data store for storing and retrieving blocks.
> * A *session* orchestrates the flow of messages with other peers for discovering and retrieving the individual blocks which comprise a single high-level object.

Reviewer 3: There is existing meaning for "block" already, better to choose a different name to avoid confusing. For instance, "blob"
```suggestion
* A *blob* represents a raw chunk of data.
* Each blob is uniquely identified by a [*content identifier (CID)*](https://docs.ipfs.io/concepts/content-addressing/), which is based on a cryptographic hash of the blob contents.
* A *blobstore* is a data store for storing and retrieving blocks.
* A *session* orchestrates the flow of messages with other peers for discovering and retrieving the individual blobs which comprise a single high-level object.
```

The full details of the Bitswap protocol can be found [here](https://research.protocol.ai/publications/accelerating-content-routing-with-bitswap-a-multi-path-file-transfer-protocol-in-ipfs-and-filecoin/delarocha2021.pdf).

> > To avoid conflicting terminology, we will henceforth refer to blocks in Bitswap as *blobs*.

Reviewer 2: ❤️ 

formatting question: why is this a quote?

Author: Here I am using this formatting as more of a note: https://stackoverflow.com/a/43120795

Probably, it would make more sense if I added the "*_NOTE:_**" in the beginning though.

### State Computation

Execution nodes, upon executing the transactions in each block, will compute the following data:

```protobuf
message ExecutionData {
    repeated Collection collections = 1;
    repeated TransactionResult txResults = 2;
    repeated Event events = 3;
    repeated TrieUpdate trieUpdates = 4;
}
```

They will then serialize each of these fields separately<sup>[1](#footnotes)</sup> and compute a root CID from the serialized data for each field as follows:

1. Split the data into blobs of up to 1MB (the message size limit of the [underlying networking stack](https://docs.ipfs.io/concepts/libp2p/)).
2. Compute the CID of each blob and store it in the blobstore.
3. If the number of CIDs computed in the previous step is 1, then return this CID. Otherwise, concatenate the CIDs and repeat from step 1<sup>[2](#footnotes)</sup>.

**[INSERT DIAGRAM HERE]**

> At the end of this process, we are left with four root CIDs (one for each field in the Execution Data). These will be included in an [Execution Result](https://github.com/onflow/flow-go/blob/master/model/flow/execution_result.go) from which an [Execution Receipt](https://github.com/onflow/flow-go/blob/master/model/flow/execution_receipt.go) is generated and later broadcast to the network.

Reviewer 2: `Execution Results` _must_ be deterministic. If two ENs have the same result, they must generate the same `Execution Result`. This implies we need protocol definitions how elements `repeated TrieUpdate trieUpdates` are ordered.

Author: I think that *if* we end up deciding on having AN's be staked, and therefore the state sync protocol becomes part of the actual core protocol, then it's also possible to have Access Nodes do the verification, as described below in the Verification section.

On the other hand, if not, then yes verifier nodes will need to do the checking, and I have written a little bit on that in the Verification section as well. 

The high level idea is that instead of generating a single `ExecutionData` per block and creating a root CID(s) for that single `ExecutionData`, the Execution nodes would instead create a separate `ExecutionData` for each chunk. Therefore, instead of just one set of root CID's in the ExecutionResult, there would be `numberOfChunks` sets of root CID's. 

Then, when a verification node is verifying each chunk, they would recompute the `ExecutionData` for that chunk, and compare the root CID's they get vs the ExecutionResult's root CID's.

@AlexHentschel does this make sense? I will add this to the doc, if there are any suggestions on how to phrase this to make it more clear please let me know.

Author: > This implies we need protocol definitions how elements repeated TrieUpdate trieUpdates are ordered.

Yes, thanks for that flag. My assumption was that there is already a defacto ordering (namely, the order that the updates were applied when executing the block), but maybe this is incorrect. Can you clarify here?

Once this is clear, I will add more detail to the doc.

Reviewer 2: > The high level idea is that instead of generating a single `ExecutionData` per block and creating a root CID(s) for that single `ExecutionData`, the Execution nodes would instead create a separate ExecutionData for each chunk.

👍  makes sense

>  defacto ordering (namely, the order that the updates occurred when executing the block)

The convention is that read and writes are deterministically ordered by their respective register ID. The ordering during execution is _part_ of the entropy that goes into creating the SPoCK proof. Therefore, we don't want to expose it as part of the result.

> At the end of this process, we are left with four root CIDs (one for each field in the Execution Data). These will be included in an [Execution Result](https://github.com/onflow/flow-go/blob/master/model/flow/execution_result.go) from which an [Execution Receipt](https://github.com/onflow/flow-go/blob/master/model/flow/execution_receipt.go) is generated and later broadcast to the network.

Reviewer 3: We serialize each field and make them as a CID, so in the end there will be 4 CIDs, since there are 4 fields. But if a field is bigger than 1MB, which exceeds the networking limit, we will further split it into smaller CIDs. So in the end, we could have more than 4 CIDs. right?

In this case, when I receive 5 CIDs in total, how do I know which 2 CIDs is more the field that is over 1MB and need to be combined?

```suggestion
At the end of this process, we are left with four root CIDs (one for each field in the Execution Data, or more if one field is bigger than 1MB). These will be included in an [Execution Result](https://github.com/onflow/flow-go/blob/master/model/flow/execution_result.go) from which an [Execution Receipt](https://github.com/onflow/flow-go/blob/master/model/flow/execution_receipt.go) is generated and later broadcast to the network.
```

Reviewer 3: Also, if a field is bigger than 1MB, how does the execution node let the access node know that there are 5 CIDs rather than 4 in total to be downloaded?

Author: @zhangchiqing 

No, we will only have a single CID, because the CID's are arranged in a tree. 

I added a diagram in the latest version, does that help clarify things?

I would also love input on how to improve the diagram :)

  <dt>Transaction Results</dt>
  <dd>Transaction Results currently do not have on-chain commitments, but these will be added in the future, allowing the downloaded Transaction Results to be verified.</dd>
</dl>

If an Access node finds that any of the downloaded objects do not have the correct hash, then the Execution node must have put an incorrect CID in their Execution Result and should be slashed.

> There is a Verifier's Dilemma here which needs to be addressed. We will not go into the details of how to solve this here.

#### Verification Nodes as Verifiers

Since Verification nodes are each responsible for only a subset of the Chunks within a block, Execution nodes will need to compute Execution Data and CIDs for each individual Chunk. Once Verification nodes have the [Chunk Data Pack](https://github.com/onflow/flow-go/blob/e73e19a87860f470053b6a8b624d3dcd43ad23bf/model/flow/chunk.go#L40), they can reproduce the CIDs and check that they are correct.

## Requirements and Rewards

> The Access node goes from being an optional node serving as a proxy, to being a relayer of state information. It should correspondingly offer SLAs on state re-transmissions and get rewarded for it.

Reviewer: I'd say the access node always had some aspects of "relaying" info, just use to be more protocol state related (and some misc data like events and result strings)

Also, as mentioned earlier, I believe discussions of the "optionality" of access nodes is still under discussion, so not sure if we want to include the idea here.

Author: yes, I should clarify and put "being a relayer of execution state information" here

Access nodes can periodically (e.g. at every 10000th block) generate a snapshot of the full state index and store it in their blobstore. They can then prune all of the Execution Data for blocks older than the snapshot, and require nodes that are behind to directly sync the snapshot first before catching up via per-block Execution Data updates.

The root CID of the snapshot must be shared with new nodes somehow so that they can use it to download the snapshot via Bitswap. There are a few ways this can be done:

* The CID can be shared on a broadcast channel using a *Byzantine consistent broadcast* algorithm<sup>[5](#footnotes)</sup>.
* A new Access node can randomly select a couple of other Access nodes and simply ask them for the CID. If the responses it receives are concordant, it can go ahead and download the data. Once the download is complete, it can generate the state trie for each block and compare its hash to the one found on-chain in the corresponding Execution Result. If it finds an inconsistency, it has all the evidence it needs to submit a slashing challenge against the nodes which provided it with the bad CID.

### Sharding

> The Execution Data can be sharded so that each Access node only needs to store and respond to requests for a fraction of the entire dataset.

Reviewer: somewhat similar to collection nodes and their clustering.

but yes, as mentioned later, this increased the severity of possible data availability issues, and should likely only be pursued after a critical mass of access nodes are part of the network
